Getting 'StructuredQuery.from cannot have more than one collection selector' from reactive repositories

Spring boot parent version  - 2.3.2.RELEASE
Spring cloud version - Hoxton.SR6

When running load test on my application getting below exception quite frequently.

`org.springframework.cloud.gcp.data.firestore.util.ObservableReactiveUtil$StreamingObserver.onError(ObservableReactiveUtil.java:126) at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:453) at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426) at io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66) at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:689) at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900(ClientCallImpl.java:577) at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:751) at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:740) at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123) at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) at java.base/java.lang.Thread.run(Unknown Source) Caused by: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: StructuredQuery.from cannot have more than one collection selector. at io.grpc.Status.asRuntimeException(Status.java:533)`

**INVALID_ARGUMENT: StructuredQuery.from cannot have more than one collection selector. at io.grpc.Status.asRuntimeException(Status.java:533)**

I am not querying on multiple collections in a single query, in-fact I am extending FirestoreReactiveRepository in it's plain form for my collections and writing methods in it for query (just like JPA). I am getting these errors for 3 repositories in my 2 Spring boot services, so I assume it won't be problem with just one of my collection. This happens when there is huge load on service, running small bunch of requests work fine.

--- a/spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/FirestoreTemplate.java
+++ b/spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/FirestoreTemplate.java
@@ -328,20 +328,20 @@ public class FirestoreTemplate implements FirestoreReactiveOperations {
 		return Mono.subscriberContext().flatMapMany(ctx -> {
 			FirestorePersistentEntity<?> persistentEntity = this.mappingContext.getPersistentEntity(clazz);
 
-			StructuredQuery.Builder builder = queryBuilder != null ? queryBuilder : StructuredQuery.newBuilder();
+			StructuredQuery.Builder builder = queryBuilder != null ? queryBuilder.clone() : StructuredQuery.newBuilder();
 			builder.addFrom(StructuredQuery.CollectionSelector.newBuilder()
 					.setCollectionId(persistentEntity.collectionName()).build());
 			if (projection != null) {
 				builder.setSelect(projection);
 			}
-			RunQueryRequest.Builder buider = RunQueryRequest.newBuilder()
+			RunQueryRequest.Builder requestBuilder = RunQueryRequest.newBuilder()
 					.setParent(this.parent)
 					.setStructuredQuery(builder.build());
 
-			doIfTransaction(ctx, resourceHolder -> buider.setTransaction(resourceHolder.getTransactionId()));
+			doIfTransaction(ctx, resourceHolder -> requestBuilder.setTransaction(resourceHolder.getTransactionId()));
 
 			return ObservableReactiveUtil
-					.<RunQueryResponse>streamingCall(obs -> this.firestore.runQuery(buider.build(), obs))
+					.<RunQueryResponse>streamingCall(obs -> this.firestore.runQuery(requestBuilder.build(), obs))
 					.filter(RunQueryResponse::hasDocument)
 					.map(RunQueryResponse::getDocument);
 		});

--- a/spring-cloud-gcp-data-firestore/src/test/java/org/springframework/cloud/gcp/data/firestore/it/FirestoreRepositoryIntegrationTests.java
+++ b/spring-cloud-gcp-data-firestore/src/test/java/org/springframework/cloud/gcp/data/firestore/it/FirestoreRepositoryIntegrationTests.java
@@ -308,4 +308,19 @@ public class FirestoreRepositoryIntegrationTests {
 		assertThat(this.userRepository.findAll().map(User::getAge).collectList().block())
 				.containsExactlyInAnyOrder(28, 59);
 	}
+
+	@Test
+	public void testDoubleSub() {
+		User alice = new User("Alice", 29);
+		User bob = new User("Bob", 60);
+		this.userRepository.save(alice).then(this.userRepository.save(bob)).block();
+
+		Mono<User> aUser = this.userRepository.findByAge(29).next();
+
+		Flux<String> stringFlux = userRepository.findAll()
+				.flatMap(user ->
+						aUser.flatMap(user1 -> Mono.just(user.getName() + " " + user1.getName())));
+		List<String> list = stringFlux.collectList().block();
+		assertThat(list).contains("Alice Alice", "Bob Alice");
+	}
 }

